framework/include/constraints/MortarSegmentHelper.h

```diff
@@ -0,0 +1,154 @@
+   * Checks whether polygons are disjoint for an easy out
+   */
+  bool isDisjoint(const std::vector<Point> & poly) const;
+
+  /**
+   * Clip secondary element (defined in instantiation) against given primary polygon
+   * result is a set of 2D nodes defining clipped polygon
+   */
+  std::vector<Point> clipPoly(const std::vector<Point> & primary_nodes) const;
+
+  /**
+   * Prints python plotting commands (to terminal) for visualizing polygons (for debugging only)
+   */
+  void plotPoly(const std::vector<Point> & poly) const;
+
+  /**
+   * Prints python plotting commands (to terminal) for visualizing triangulation (for debugging
+   * only)
+   */
+  void plotTriangulation(const std::vector<Point> & nodes,
+                         const std::vector<std::vector<unsigned int>> & elem_to_nodes) const;
+
+  /**
+   * Triangulate a polygon (currently uses center of polygon to define triangulation)
+   *
+   * Inputs:
+   * - List of 2D nodes defining polygon
+   *
+   * Outputs:
+   * - List of integer arrays defining which nodes belong to each triangle
+   * - For polygons with more than 3 nodes, appends center point to nodes
+   */
+  void triangulatePoly(std::vector<Point> & nodes,
+                       const unsigned int offset,
+                       std::vector<std::vector<unsigned int>> & tri_map) const;
+
+  /**
+   * Get mortar segments generated by a secondary and primary node pair
+   */
+  void getMortarSegments(const std::vector<Point> & primary_nodes,
+                         std::vector<Point> & nodes,
+                         std::vector<std::vector<unsigned int>> & elem_to_nodes);
+
+  /**
+   * Compute area of polygon
+   */
+  Real area(const std::vector<Point> & nodes) const;
+
+  /**
+   * Get center point of secondary element
+   */
+  const Point center() { return _center; }
+
+  /**
+   * Get area fraction remaining after clipping against primary elements
+   */
+  Real remainder() const { return _remaining_area_fraction; }
+
+  /**
+   * Get 3D position of node of linearized secondary element
+   */
+  const Point point(unsigned int i) const
+  {
+    return (_secondary_poly[i](0) * _u) + (_secondary_poly[i](1) * _v) + _center;
+  }
+
+private:
+  /**
+   * Geometric center of secondary element
+   */
+  Point _center;
+
+  /**
+   * Normal at geometric center of secondary element
+   */
+  Point _normal;
+
+  /**
+   * Vectors orthogonal to normal that span the plane projection will be performed on.
+   * These vectors are used to project the polygon clipping problem on a 2D plane,
+   * they are defined so the nodes of the projected polygon are listed with positive orientation
+   */
+  Point _u, _v;
+
+  /**
+   * Area of projected secondary element
+   */
+  Real _secondary_area;
+
+  /**
+   * Fraction of area remaining after overlapping primary polygons clipped
+   */
+  Real _remaining_area_fraction;
+
+  bool _debug;
+
+  /**
+   * Tolerance for intersection and clipping
+   */
+  Real _tolerance = 1e-8;
```

Comment:
I don't think so, it really shouldn't be changed (it's not as delicate now that I've modified how the intersection is computed). It's used in a few places like to help with the conditioning of the getIntersection routine (which is horribly conditioned when liines are parallel like when dealing with conforming meshes). Before I put this `s` business into the calculation, when the problem was poorly conditioned it would spit out random points that weren't even on my surfaces. The tolerancing is less important for the intersection now that it's impossible for it to return points not on the polygon but it also plays a role in limiting the proliferation of nodes (before it would produce about 4x as many nodes as needed for conforming meshes which in turn created a bunch of tiny mortar segments). 

It works well now, I haven't any problems and have tested it thoroughly but this clipping stuff is one part of the mortar code I would avoid modifying, it looks benign but it took multiple iterations to get working robustly.